### ISSUE TITLE
Fix S1854 FP: Try/Catch statement with constant value propagation

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-dotnet#2596]
In the following code, SonarAnalyzer.CSharp 7.16.0.8981 flags `S1854: Remove this useless assignment to local variable 'length'.` Is this a false positive?

```
using System;
using System.IO;

namespace S1854_False_Positive
{
    public static class C
    {
        public static long GetFileLength(string path)
        {
            const int unknownfilelength = -1;
            long length = unknownfilelength;
            try
            {
                length = new FileInfo(path).Length;
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
            return length;
        }
    }
}
```

**Comment by antoinebj** [comment]
I also experienced this false positive on that rule since 7.16.
In my case the variable was initialized as `(string)null` and strangely, it seems that changing to `default(string)` made the FP not happen anymore.

**Comment by andrei-epure-sonarsource** [comment]
thanks @nnpcYvIVl - this is because we're not following constant value propagation (yet). If you replace that with `long length = -1;` , we won't raise, because `-1` is usually a default value. To fix your FP, we'd have to enhance the engine with the ability to track constants across variables, and we don't have plans for that in the immediate future. I've added a test case with the FP to the rule, though.

thanks @antoinebj - if you initialize `string s = null` it won't raise. There is a bug here because we don't support the cast `(string) null`

**Comment by andrei-epure-sonarsource** [comment]
I've split the ticket in 2 because there's 2 different issues here. See #2598
